File: C++/Arrays/MaximumSubarraySum.cpp
//Optimal Solution
// Time Complexity: O(n)
// Space Complexity: O(1)
//algo used Kadane Algorithm
class Solution{
    public:
        int maxSubArray(vector<int>&nums){
            int cursum=0;
            int maxsum=INT_MIN;
            for(auto num:nums){
                cursum=max(num,cursum+num); //find which subarray has max
                maxsum=max(maxsum,cursum); //max sum
            }
        }
        return maxsum;
}

File: JAVA/Arrays/MaximumSubArraySum.java
//Optimal Solution
// Time Complexity: O(n)
// Space Complexity: O(1)
//algo used Kadane Algorithm
public class MaximumSubArraySum {
    public int maxSubArray(int[] nums) {
        int maxSum = nums[0];
        int CurrSum = nums[0];
        for (int i = 1; i < nums.length; i++) {
            CurrSum = Math.max(nums[i], CurrSum + nums[i]);
            maxSum = Math.max(maxSum, CurrSum);
        }
        return maxSum;
    }
}

